### ISSUE TITLE
Session feature fails with OOM error in 0.32.1

### ISSUE THREAD

**Issue by adamreichold** [rusqlite/rusqlite#1590]
When upgrading [an application](https://github.com/adamreichold/b2_backup) from version 0.31.0 to 0.32.1, `Changegroup::add_stream` started consistently failing with

```
SqliteFailure(Error { code: OutOfMemory, extended_code: 7 }, None)
```

Downgrading back to 0.31.0 appears to work around the issue.

I am using

```toml
rusqlite = { version = "0.32", features = ["bundled", "session"] }
```

to enable access to the session feature. The specific code in question is [here](https://github.com/adamreichold/b2_backup/blob/3f2af2d041892af5bcfd4011d467efb7ae372d92/src/manifest.rs#L260).

**Comment by gwenn** [comment]
Hard to say !
```sh
git log -p v0.31.0..v0.32.1 src/session.rs
```
?
Bundled SQLite 3.45.1 vs 3.46.0 ?
Or another leak somewhere ?
Could you please try to reproduce the OOM with `rusqlite` v0.32.1 and SQLite 3.45.1 (without `bundled`) ?

**Comment by adamreichold** [comment]
> Could you please try to reproduce the OOM with rusqlite v0.32.1 and SQLite 3.45.1 (without bundled) ?

Do you have any recommendation on how to best approach this? My distribution also ships v3.46.0.

Could I just checkout an earlier version of the `libsqlite3-sys/sqlite3` folder in a clone of this repository and continue to use bundled?

**Comment by gwenn** [comment]
You can also try `SQLITE3_LIB_DIR` / `SQLITE3_INCLUDE_DIR`.
Or try to run your application with ASAN enabled like here:
https://github.com/rusqlite/rusqlite/blob/67cb6f1576087bf7674f1d4ca67864aba96f53fb/.github/workflows/main.yml#L158

**Comment by adamreichold** [comment]
> Or another leak somewhere ?

By the way, I don't think there actually is an out of memory issue. The application was happily running in a retry loop in the background due to this without the system experiencing memory pressure. I think this rather sounds like type confusion or memory corruption.

**Comment by adamreichold** [comment]
Indeed, v0.32.1 with only c771c222cdee4e6262ea31c5013b37626273582c merged also works, so I suspect that the bump to 3.46.0 somehow went awry.

**Comment by gwenn** [comment]
Maybe related to #585 impacting "stream".
See https://github.com/rusqlite/rusqlite/issues/585#issuecomment-549028440

**Comment by gwenn** [comment]
See also #764 ...

**Comment by adamreichold** [comment]
Hhhmmm, so enabling the `buildtime_bindgen` feature on v0.32.1 does not help, even after upgrading bindgen to 0.70.1 as on master, so the Rust bindings do not seem to be the issue...

> Maybe related to https://github.com/rusqlite/rusqlite/issues/585 impacting "stream".

But that would not really fit that upgrading to SQLite 3.47.0 fixed the issue, would it?

**Comment by gwenn** [comment]
Only `sqlite3changeset_start_strm` should be impacted by #585:
https://sqlite.org/session.html#sqlite3changegroup_add_strm
> In the case of sqlite3changeset_start_strm(), the xInput callback may be invoked by the sessions module at any point during the lifetime of the iterator.

Not `add_stream` / `sqlite3changegroup_add_strm`.
I will try to check our `xInput` impl relative to these rules:
> Assuming no error occurs the xInput method should copy up to (*pnData) bytes of data into the buffer and set (*pnData) to the actual number of bytes copied before returning SQLITE_OK. If the input is completely exhausted, (*pnData) should be set to zero to indicate this.

**Comment by gwenn** [comment]
https://github.com/sqlite/sqlite/commit/253695caf33a6cbf8c02e8af1e2f2a7e57ec04a0
?

**Comment by adamreichold** [comment]
Yes, that looks like a likely culprit. So I'll just wait for the next release of rusqlite which includes the fixed SQLite 3.47. Should this be kept open until then for anybody else running into this?

**Comment by gwenn** [comment]
Yep, no hurry to close this issue.

**Comment by adamreichold** [comment]
Closing as there have been two releases of rusqlite in the meantime.
